### PR TITLE
feat(nix): improve `flake.nix`, adding `checks` and `formatter`

### PR DESCRIPTION
**Pull request by johnyob**
# Context

**Dependencies**: #543 
**Related Tasks**: [jstz-45](https://linear.app/tezos/issue/JSTZ-45/configuring-nix-for-ci), [jstz-48](https://linear.app/tezos/issue/JSTZ-48/use-rust-toolchaintoml-file-in-nix-shell-and-jstz-derivation)

This PR is the first one in the line of work of transitioning our CI to an entirely Nix-based environment. 

# Description

This PR contains a number of changes (separated into commits):
- Use `crane` to build our Rust crates -- `crane` is a Nix library for building Rust projects with excellent integration into tooling and caching of dependencies
- Fix the broken derivations for the `@jstz-dev/jstz` and `@jstz-dev/types` packages
- Add cargo unit tests to `nix flake check`s
- Configure `nix fmt` to format the codebase

# Manually testing the PR

### CODE REVIEW

**Comment by zcabter** [comment]
Reviewing

File: nix/crates.nix [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  stdenv,
+  rust-toolchain,
+}: let
+  craneLib = (crane.mkLib pkgs).overrideToolchain (_: rust-toolchain);
+
+  # TODO(https://linear.app/tezos/issue/JSTZ-68):
+  # Filter crate srcs into a file set to avoid rebuilding all
+  # crates when a single crate changes.
+  src = let
+    # Include all Rust files / cargo releated files
```
Comment (zcabter) [inline nix/crates.nix:14]:
```suggestion
    # Include all Rust files / cargo related files
```

**Comment by zcabter** [comment]
Is it a good idea to include `nix fmt` in the pre-commit hook?

**Comment by zcabter** [comment]
Build and tests runs locally.

File: flake.nix [hunk 2]
```diff
@@ -49,22 +55,34 @@
...
+
+            # TODO(https://linear.app/tezos/issue/JSTZ-63)
+            # Configure formatter for LIGO contracts
+            settings.global.excludes = ["target" "result" "node_modules/**" ".github" ".direnv" "contracts/**" "Dockerfile" "*.toml"];
           };
+        in {
+          packages = crates.packages // js-packages.packages // {default = self.packages.${system}.jstz_kernel;};
+          checks = crates.checks // {formatting = fmt.config.build.check self;};
+
+          formatter = fmt.config.build.wrapper;
```
[the remark this answers is not in the document]
Comment (johnyob) [reply flake.nix:82]: `formatter` is used to configure `nix fmt`, `checks` is used to configure `nix flake check` and `packages` is used to configure `nix build`

**Comment by johnyob** [comment]
> Maybe its a good idea to include nix fmt in the pre-commit hook?

I agree. We should be able to do this fairly easily with https://github.com/cachix/git-hooks.nix. Perhaps a different PR?

**Comment by zcabter** [comment]
> > Maybe its a good idea to include nix fmt in the pre-commit hook?
> 
> I agree. We should be able to do this fairly easily with https://github.com/cachix/git-hooks.nix. Perhaps a different PR?

Ok let's do it in a follow up. Tracked here: https://linear.app/tezos/issue/JSTZ-73/run-nix-fmt-in-pre-commit-hook

**Review by zcabter — APPROVED** [review]

**Pull request merged**